service/history/workflow/cache/cache.go

```diff
@@ -328,11 +308,13 @@ func (c *cacheImpl) getOrCreateWorkflowExecutionInternal(
 			shardContext.GetMetricsHandler(),
 		)
 
-		var err error
-		err = c.Put(shardContext, namespaceID, execution, workflowCtx, handler)
+		value := &cacheItem{shardId: shardContext.GetShardID(), wfContext: workflowCtx, finalizer: shardContext.GetFinalizer()}
+		elem, err := c.PutIfNotExist(cacheKey, value)
 		if err != nil {
+			metrics.CacheFailures.With(handler).Record(1)
 			return nil, nil, err
 		}
+		workflowCtx = elem.(*cacheItem).wfContext
```

Comment:
Reverted back to what it was before I introduced `Put`:
https://github.com/temporalio/temporal/commit/f34f22c4a3a0e0a6ec726b53b2ef212116a184b7#diff-53066639dfa160c9360f0e33b29a37460691840bd9d86413be8bc839faa49623L240-L256